### PR TITLE
Crypto: general updates

### PR DESCRIPTION
**Pull request by tarakby**
- update KMAC implementation to force keys to be at least 128 bits. In Flow, KMAC keys are only used as domain separation tags and do not need to be checked against the minimum size, but the check is enforced since the API is public.
- make sure all instances of KMAC are using keys with a valid length.
- add more checks on the threshold signature stateful and stateless api. 
- add a missing check in a feldman vss api.
- use a max limit for seed length when generating ECDSA private keys.

### CODE REVIEW

File: model/encoding/signing_tags.go [hunk 1]
```diff
@@ -12,13 +8,16 @@ import (
 // a domain tag that specifies the type of the signed object.
 
 func tag(domain string) string {
-	return protocolPrefix + domain + crypto.BLSCipherSuite
```
Comment (turbolent) [inline model/encoding/signing_tags.go:15]: Why was this removed? Is this backwards-compatible change?

Comment (tarakby) [reply model/encoding/signing_tags.go:15]: yes, it was moved here: https://github.com/onflow/flow-go/blob/d5f70f2e8e7b8f8403cc494a3a67381b1edc2f85/crypto/bls.go#L92 to be part of the crypto code and the audit scope.

Comment (tarakby) [reply model/encoding/signing_tags.go:15]: Actually, I just updated it: 
https://github.com/onflow/flow-go/blob/3d8787cc9e8dd60faa689797e9903cee47da8f65/crypto/bls.go#L92

**Review by arrivets — APPROVED** [review]

**Review by zhangchiqing — APPROVED** [review]
Looks good. have some minor comments

File: crypto/thresholdsign.go [hunk 2]
```diff
@@ -109,11 +110,16 @@ func NewThresholdSigner(size int, threshold int, currentIndex int, hashAlgo hash
...
 func (s *thresholdSigner) SetKeys(currentPrivateKey PrivateKey,
 	groupPublicKey PublicKey,
-	sharePublicKeys []PublicKey) {
+	sharePublicKeys []PublicKey) error {
+
+	if len(sharePublicKeys) != s.size {
+		return fmt.Errorf("size of public key shares should be %d", s.size)
+	}
 
 	s.currentPrivateKey = currentPrivateKey
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/thresholdsign.go:119]: That's right, same comment also applies to `SetMessage`. On the other hand, we would like to still be able to rotate the keys and message on the same threshold instance. 
Since the final threshold signature is always verified by the api, any error will be caught, but maybe we need a cleaner way to avoid errors.

File: state/protocol/seed/seed.go [hunk 3]
```diff
@@ -33,7 +33,11 @@ func FromRandomSource(indices []uint32, sor []byte) ([]byte, error) {
 	}
 
 	// create the key used for the KMAC by concatenating all indices
-	key := make([]byte, 4*len(indices))
+	keyLen := 4 * len(indices)
+	if keyLen < hash.KmacMinKeyLen {
+		keyLen = hash.KmacMinKeyLen
```
[the remark this answers is not in the document]
Comment (tarakby) [reply state/protocol/seed/seed.go:38]: Yes, correct

File: crypto/thresholdsign.go [hunk 4]
```diff
@@ -340,14 +347,22 @@ func ReconstructThresholdSignature(size int, threshold int,
...
 	flatShares := make([]byte, 0, signatureLengthBLSBLS12381*(threshold+1))
 	indexSigners := make([]index, 0, threshold+1)
 	for i, share := range shares {
 		flatShares = append(flatShares, share...)
+		// check the index is valid
 		if signers[i] >= size || signers[i] < 0 {
 			return nil, fmt.Errorf("signer index #%d is invalid", i)
 		}
+		// check the index is new
+		if _, isSeen := m[index(signers[i])]; isSeen {
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/thresholdsign.go:363]: Oups, good catch! 
This is to make the api overall safer, regardless of what the caller do.

File: crypto/thresholdsign.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/thresholdsign.go:123]: I think I could just call `ClearShares` when setting keys. This erases all existing shares that are signed with older keys.

File: crypto/thresholdsign.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/thresholdsign.go:368]: added a test to cover this case.
